pylauncher4.py

```diff
@@ -12,7 +12,10 @@
 import subprocess
 import sys
 import time
-import hostlist3 as hs
+#import hostlist3 as hs
+# hostlist3 doesn't appear to be listed on conda-forge. Is there a reason why we use it over the
+# module hostlist from the python-hostlist package?
+import hostlist as hs
 
```

Comment:
Yes there actually is a reason for this. And its quite convoluted and hacky, bother it goes:

pylauncher4.py is actually just a patch to the pylauncher python library, https://github.com/TACC/pylauncher,  which is maintained by TACC and loaded as a module that is available on TACC computers. pylauncher4.py extends pylauncher to allow for pre and post processing commands to take place before main execution of parallel jobs. hostlist3 is a library that is available in the pylauncher library. So when the module is loaded on TACC computers hostlist3 should be available. See code below:

```
(base) login4.stampede2(1002)$ python
Python 3.8.3 (default, May 19 2020, 18:47:26)
[GCC 7.3.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import hostlist3
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'hostlist3'
>>> exit()
(base) login4.stampede2(1003)$ module load pylauncher

Lmod is automatically replacing "python2/2.7.15" with "python3/3.7.0".

(base) login4.stampede2(1004)$ python
Python 3.8.3 (default, May 19 2020, 18:47:26)
[GCC 7.3.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import hostlist3
>>>
```

So APS is dependent upon the pylauncher library, which isn’t part of the Conda installation, but rather a module loaded by TACC. And in fact, different components of the APS application run using different python environments. The main entry-point of the application, https://github.com/cdelcastillo21/APS/blob/main/wrapper.sh, shows this in action.